File: packages/tep-kotlin/tep/src/main/kotlin/uz/asad/tep/Signature.kt
package uz.asad.tep

import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.util.Base64
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

object Signature {
    const val SIG_PREFIX = "v1."

    fun canonical(envelope: TepEnvelope, rawPayload: ByteArray): ByteArray {
        val joined = listOf(
            "tep",
            TepEnvelope.VERSION,
            envelope.eventId,
            envelope.timestamp,
            envelope.source,
            envelope.type,
            String(rawPayload, StandardCharsets.UTF_8),
        ).joinToString("\n")
        return joined.toByteArray(StandardCharsets.UTF_8)
    }

    fun sign(envelope: TepEnvelope, rawPayload: ByteArray, secret: ByteArray): String {
        return SIG_PREFIX + Base64.getUrlEncoder().withoutPadding().encodeToString(rawHmac(envelope, rawPayload, secret))
    }

    fun verify(envelope: TepEnvelope, rawPayload: ByteArray, signature: String, secret: ByteArray): Boolean {
        if (!signature.startsWith(SIG_PREFIX)) return false
        val provided = try {
            Base64.getUrlDecoder().decode(signature.removePrefix(SIG_PREFIX))
        } catch (e: IllegalArgumentException) {
            return false
        }
        val expected = rawHmac(envelope, rawPayload, secret)
        return MessageDigest.isEqual(provided, expected)
    }

    fun rawHmac(envelope: TepEnvelope, rawPayload: ByteArray, secret: ByteArray): ByteArray {
        require(secret.size >= 32) { "secret must be at least 32 bytes" }
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(secret, "HmacSHA256"))
        return mac.doFinal(canonical(envelope, rawPayload))
    }
}

File: packages/tep-typescript/src/signature.ts
import { createHmac, timingSafeEqual } from 'node:crypto';
import type { TepEnvelope } from './types.js';
import { VERSION } from './types.js';

const SIG_PREFIX = 'v1.';

export function canonical(envelope: TepEnvelope, rawBody: Buffer): Buffer {
  const parts = [
    'tep',
    VERSION,
    envelope.event_id,
    envelope.timestamp,
    envelope.source,
    envelope.type,
    rawBody.toString('utf8'),
  ];
  return Buffer.from(parts.join('\n'), 'utf8');
}

export function sign(envelope: TepEnvelope, rawBody: Buffer, secret: string): string {
  return SIG_PREFIX + createHmac('sha256', secret).update(canonical(envelope, rawBody)).digest('base64url');
}

export function verify(envelope: TepEnvelope, rawBody: Buffer, signature: string, secret: string): boolean {
  try {
    if (typeof signature !== 'string' || !signature.startsWith(SIG_PREFIX)) {
      return false;
    }
    const provided = Buffer.from(signature.slice(SIG_PREFIX.length), 'base64url');
    const expected = createHmac('sha256', secret).update(canonical(envelope, rawBody)).digest();
    if (provided.length !== expected.length) {
      return false;
    }
    return timingSafeEqual(provided, expected);
  } catch {
    return false;
  }
}